File: experiments/game-of-life/csharp/attempts/opus46high/5/GameOfLife.cs
using System.Collections.Generic;

namespace GameOfLife;

public enum Cell { Dead, Alive, Zombie }

public record Position(int Row, int Col);

public record Grid(int Width, int Height, Dictionary<Position, Cell> Cells)
{
    public Cell? CellAt(Position position)
    {
        return Cells.TryGetValue(position, out var cell) ? cell : null;
    }

    public int CountLiveNeighbors(Position position)
    {
        (int dr, int dc)[] offsets =
        [
            (-1, -1), (-1, 0), (-1, 1),
            (0, -1),           (0, 1),
            (1, -1),  (1, 0),  (1, 1)
        ];

        var count = 0;
        for (var i = 0; i < offsets.Length; i++)
        {
            var neighbor = new Position(position.Row + offsets[i].dr, position.Col + offsets[i].dc);
            var c = CellAt(neighbor);
            if (c == Cell.Alive || c == Cell.Zombie)
                count++;
        }
        return count;
    }

    public static Cell EvolveCell(Cell cell, int liveNeighbors) => (cell, liveNeighbors) switch
    {
        (Cell.Zombie, _) => Cell.Zombie,
        (Cell.Alive, 2) => Cell.Alive,
        (Cell.Alive, 3) => Cell.Alive,
        (Cell.Dead, 3) => Cell.Alive,
        _ => Cell.Dead
    };

    public Grid Tick()
    {
        var newCells = new Dictionary<Position, Cell>();
        for (var row = 0; row < Height; row++)
        {
            for (var col = 0; col < Width; col++)
            {
                var pos = new Position(row, col);
                var current = CellAt(pos) ?? Cell.Dead;
                var neighbors = CountLiveNeighbors(pos);
                newCells[pos] = EvolveCell(current, neighbors);
            }
        }
        return this with { Cells = newCells };
    }
}

File: experiments/game-of-life/fsharp/attempts/codex54high/3/GameOfLife.fs
module GameOfLife

type Cell =
    | Alive
    | Dead
    | Zombie

type Position = { Row: int; Col: int }

type Grid =
    { Width: int
      Height: int
      Cells: Map<Position, Cell> }

let cellAt (grid: Grid) (position: Position) : Cell option = Map.tryFind position grid.Cells

let isLiveNeighbor cell =
    match cell with
    | Alive
    | Zombie -> true
    | Dead -> false

let countLiveNeighbors (grid: Grid) (position: Position) : int =
    [ (-1, -1)
      (-1, 0)
      (-1, 1)
      (0, -1)
      (0, 1)
      (1, -1)
      (1, 0)
      (1, 1) ]
    |> List.map (fun (dr, dc) ->
        cellAt
            grid
            { Row = position.Row + dr
              Col = position.Col + dc })
    |> List.filter (Option.exists isLiveNeighbor)
    |> List.length

let evolveCell (cell: Cell) (liveNeighbors: int) : Cell =
    match cell, liveNeighbors with
    | Zombie, _ -> Zombie
    | Alive, 2
    | Alive, 3 -> Alive
    | Dead, 3 -> Alive
    | _ -> Dead

let tickGrid (grid: Grid) : Grid =
    let cells: Map<Position, Cell> =
        [ for row in 0 .. grid.Height - 1 do
              for col in 0 .. grid.Width - 1 do
                  let pos = { Row = row; Col = col }
                  let current = cellAt grid pos |> Option.defaultValue Dead
                  let neighbors = countLiveNeighbors grid pos
                  pos, evolveCell current neighbors ]
        |> Map.ofList

    { grid with Cells = cells }
